[sw, dif_pinmux] Introduce DIF Pinmux library with unit tests

silvestrst opened the pull request:
This PR:
- Introduces DIF Pinmux library.
- Introduces DIF Pinmux library unit tests.
- Change the current Pinmux "driver" to use the DIF Pinmux.

NOTE:
This DIF is not complete - the corresponding IP is being extended to include sleep/wake-up and other functionality. However, adding new features to the Pinmux IP does not affect the existing registers and functionality (**meaning that the current functionality will not need to be changed, only extended**).

silvestrst commented:
Just realised that I forgot to run clang-format, will push an update shortly.

silvestrst commented on sw/device/lib/base/mmio.h:
```diff
@@ -234,6 +234,30 @@ inline void mmio_region_nonatomic_set_mask32(mmio_region_t base,
+                                              ptrdiff_t offset,
+                                              uint32_t mask,
+                                              uint32_t index,
+                                              uint32_t value) {
```
Index and value could be passed inside a structure, that probably would align the API better with potential @mcy suggested changes to the reggen, see #1710 .

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,378 @@
+TEST_F(DifPinmuxTest, InselSetFailureBadMioSelect) {
+  DeviceInit();
+
+  int less_than_min = (int)kDifPinmuxInselTieToZero - 1;
```
Prefer
```cc
auto foo = static_cast<int>(...);
```

In general, avoid C-like casts like the plague in C++ and use `static_cast` instead.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,378 @@
+      PINMUX_PERIPH_INSEL0_REG_OFFSET,
+      {
+          {PINMUX_PERIPH_INSEL0_IN0_OFFSET, PINMUX_PERIPH_INSEL0_IN0_MASK,
+           (uint32_t)kDifPinmuxInselTieToZero},
```
I can never remember... does C++ really not define an implicit conversion for this? Yuck.

Could always just do
```
#ifdef __cplusplus
#define ENUM32(foo) enum foo : uint32_t
#else
#define ENUM32(foo) enum foo
#endif
```
but that's kind of a bad idea.

Sigh, C++.

silvestrst replied:
I am not sure what my opinion on this is, I think normally I prefer explicit casts.

mcy commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,232 @@
+
+  // Register offset based on a given peripheral input and MIO select.
+  size_t register_index =
+      (size_t)peripheral_input / pinmux->priv->insel_fields_per_register;
```
I'd add some parentheses here to make the desired scope of the cast clear. I can never remember the correct precedence rules.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,226 @@
+ * TODO - this won't be needed when all of the information in this structure
+ *        is autogenerated.
+ */
+struct dif_pinmux_private {          // TODO - not needed when other TODOs fixed
```
Not sure that the way I went about is the right thing, but there is an incomplete type in the header. The idea is to have a named "private" field in the `dif_plic` structure, which is not accessible outside of the source file.

mcy replied:
If it's global (there's just one static), why bother sticking a pointer to in it in the pinmux struct?

(Opaque types are pretty gross in the absence of an allocator...)

silvestrst replied:
That is indeed a fair point, there is no need for a pointer in the pinmux struct. This definitely needs to be rewritten. Thanks.

silvestrst replied:
Done.

silvestrst commented:
!!!DO NOT REVIEW FOR NOW!!!

With #1835 will be half the size it is now.

silvestrst commented:
In the latest push:
* Refactored the DIF Pinmux Library to use #1835

silvestrst commented:
In the latest push:
* Have resolved all of the outstanding comments.
* Changed `size_t` for `uint32_t`.

silvestrst commented:
In the latest push:
* Have fixed the `dif_pinmux_insel_t` range (`kDifPinmuxInselPadLast = kDifPinmuxInselTieToOne + DIF_PINMUX_MIO_NUM` instead of `kDifPinmuxInselPadFirst`).
* Have removed unnecessary explicit casts from the `.cc`.
* Have introduced `UnittestsUpToDate` unittest.

silvestrst commented:
Outstanding CI failures are: python lint failure on a temporary commit, CI error (can't find ninja 1.5) - both these errors can be ignored.

silvestrst commented:
My clever idea of defining `const uint32_t kDif... = autogenerated_define` in the source file, and using it inside header file, by declaring `extern const uint32_t kDif...` didn't work. I wanted to do something like that:
```
typedef enum dif_pinmux_insel {
  kDifPinmuxInselTieToZero = 0,
  kDifPinmuxInselTieToOne = 1,
  kDifPinmuxInselPadFirst = 2,
  kDifPinmuxInselPadLast = kDifPinmuxInselPadLast + kDif...,
} dif_pinmux_insel_t
```
But then you rightly get the following errors:
```
error: the value of ‘...’ is not usable in a constant expression
error: enumerator value for ‘...’ is not an integer constant
```

I don't want to use the `pinmux_regs.h` in the header file, because it means that any user of the dif needs to link against the generated defines.

silvestrst commented:
In the latest push:
* I have rebased pinmux library on top of the upstream.

gkelly commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+#include "pinmux_regs.h"  // Generated.
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - autogenerate?
```
If these aren't generated then I'd file an issue and remove the TODO here, it will just rot over time. I also feel like `REGISTER_WIDTH` isn't a great description because it's more than one register in this bundle, right? Maybe `REGISTER_BLOCK_WIDTH`? Something to indicate that it's more than one register.

silvestrst replied:
Sure, I will fire an issue. I think both of the things you have mentioned in this comment are subject of a separate issue/PR, as this will include changes in reggen that will also affect other peripherals.

lenary replied:
Don't file an issue if this is covered by in-progress PRs (which I think it might be?)

silvestrst replied:
I think it should have been a part of #1814 but was missed. I have opened an issue and marked it as a good first issue, I don't think it is that important for the time being anyway.

silvestrst replied:
All TODOS have been removed in `.{c,h}`.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+// TODO - autogenerate?
+#define PINMUX_REGISTER_WIDTH_IN_BYTES (PINMUX_PARAM_REG_WIDTH / 8)
+
+// TODO - autogenerate?
```
Same here.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+  pinmux->base_addr = base_addr;
+  pinmux->write_enabled = write_enabled;
+
+  // Set the peripheral to the reset defaults.
```
The comment here and on `dif_pinmux_reset` are repetitious, the function is pretty clearly named so I'd just drop them.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+    return kDifPinmuxInselSetErrorInvalidArgument;
+  }
+
+  if (!pinmux->write_enabled) {
```
This seems like it's just duplicating the `PINMUX_REGEN_WEN` bit, why not read that instead of storing state?

silvestrst replied:
I had this conversation with @lenary , kind of seemed that it is a bit wasteful to read this register on every function call if the value is known.

I can change it, but then it would result in a volatile read every function call.

lenary replied:
We briefly discussed this and my view is that it's ok if this function *attempts* to change the `insel`, and disregards the possibility that this might fail because writes were disabled - ie the function always performs the write, and always assumes it succeeded. This means you don't need to track `write_enabled`, and nor do you need to read the WEN register in every function (and you no longer need `kDifPinmuxInselSetErrorRegistersLocked`).

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,231 @@
+
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - should be autogenerated?
+#define DIF_PINMUX_MIO_NUM 32
```
I'd file an issue and drop this comment. The name is also unclear to me, is this the number of MIOs? The count? The "number" in some other sense? It seems like it's not used very much, do DIF users need to know anything about it?

silvestrst replied:
I think you are right, I don't need to expose it to the outside world. However, before we have auto generated the enums for `dif_pinmux` in `top_earlgrey` (similar to what we have done for plic IRQs), I will have to expose `PINMUX_PARAM_NMIOPERIPHIN`, `PINMUX_PARAM_NMIOPERIPHOUT`, and couple more defines.

silvestrst replied:
This will be refactored, for now we probably still need to expose some of the defines, but eventually the corresponding enums will be generated similar to plic interrupts.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,231 @@
+  kDifPinmuxOutselOutGpio29,
+  kDifPinmuxOutselOutGpio30,
+  kDifPinmuxOutselOutGpio31,
+  kDifPinmuxOutselOutLast = kDifPinmuxOutselOutGpio31,
+} dif_pinmux_outsel_t;
+#define DIF_PINMUX_OUTSEL_NUM (kDifPinmuxOutselOutLast + 1)
```
This seems like duplication, why have two ways of determining the last value? I thought we'd standardized on one in an earlier DIF review.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,231 @@
+ * Initialises an instance of Pinmux.
+ *
+ * Information that must be retained, and is required to program Pinmux, shall
+ * be stored in @p pinmux.
```
`` ` ``-style parameter names, here and below.

silvestrst replied:
Done.

gkelly commented on sw/device/lib/pinmux.c:
```diff
@@ -6,36 +6,40 @@
-                  32 / PINMUX_PERIPH_INSEL0_IN1_OFFSET,
-                  PINMUX_PERIPH_INSEL0_IN0_MASK,
-                  PINMUX_PERIPH_INSEL_IDX_OFFSET);
+  // Assign a MIO outputs to peripherl inputs (1:1, don't assign if peripheral
```
nit: peripheral

silvestrst replied:
Done.

gkelly commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,379 @@
+
+extern "C" {
+#include "pinmux_regs.h"  // Generated.
+#include "sw/device/lib/dif/dif_pinmux.h"
```
Should we be inserting `extern "C"` in our headers I wonder?

mcy replied:
I think so. Wrapping includes in extern "C" is kind of rude IMO; it's probably good practice to mark polyglot headers as such with the usual `__cplusplus` incantation. I'll file an issue, but I don't think we should worry about it in this change.

https://github.com/lowRISC/opentitan/issues/1934

silvestrst replied:
We still have to wrap "pinmux_regs.h" as they are not auto generated with
```
#ifdef __cplusplus
extern "C" {
#endif  // __cplusplus 
```

So I am leaving this as it is for now.

lenary replied:
Sounds good, We need to look at the fall-out from the polyglot headers rules anyway. :)

lenary replied:
#2308 is ready for the generated headers, but I don't know when it will be merged. I'm happy for this to be left as-is and cleared up when we make all the DIF headers polyglot (which we probably want to do all at once).

gkelly commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,379 @@
+        PINMUX_MIO_OUTSEL6_REG_OFFSET,
+    };
+    for (const auto &reg : outsel_registers) {
+      // TODO - autogenerate?
```
Is this autogenerateable? I'm not sure given what our hjsons currently carry. I'd drop the comment.

silvestrst replied:
Done.

gkelly commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,379 @@
+// Tests to ensure that this unittest is up-to-date with the dif_pinmux library.
+
+TEST_F(DifPinmuxTest, UnittestsUpToDate) {
+  EXPECT_EQ(DIF_PINMUX_MIO_NUM, PINMUX_PARAM_NMIOPADS);
```
This test isn't testing whether the unit test is up to date but whether the DIF matches the IP parameters. I wonder if these shouldn't be `static_assert`'d in the DIF itself. @lenary, thoughts on letting people `#include <assert.h>` and `static_assert` in a DIF?

silvestrst replied:
That would be more convenient, if there is no problem in using asserts in DIFs then im pro it.

silvestrst replied:
@lenary, what do you think?

lenary replied:
I think it makes a lot of sense to state the assumptions around parameterisation of hardware IP in a way that the compiler can catch (so using `static_assert`).

I am worried about including `<assert.h>`, because that requires an implementation of `assert` which requires an implementation of `abort`. Thankfully C11 provides a workaround: You can use `_Static_assert(cond, message)` without including assert.h (it's a keyword, like `sizeof`).

So, it makes sense to `_Static_assert(…);` any assumptions about parameterisations which can be matched against generated constants, so that we can catch these issues.  These asserts should probably be put in the `dif_<ip>.c` file, in the top-level scope, immediately after the headers are included, and should have documentation.

silvestrst replied:
Only `*FIELD_WIDTH`, `*FIELDS_PER_REG`, `*MULTIREG_COUNT` are relevant and checked via static assert.

Reset register values are relevant as well, but they are not auto generated yet, so nothing to do there.

silvestrst replied:
Done.

silvestrst commented:
@gkelly thank you for review.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,379 @@
+  EXPECT_EQ(kDifPinmuxMioInPadLast, PINMUX_PARAM_NMIOPADS - 1);
+}
+
+// dif_pinmux_init() tests.
```
Consider declaring a derived class of `DifPinmuxTest` for each suite of tests (e.g. like in my SPI diff).

silvestrst replied:
That was what I initially was doing, and then switched, ill do that.

silvestrst replied:
Done.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,379 @@
+
+// Tests to ensure that this unittest is up-to-date with the dif_pinmux library.
+
+TEST_F(DifPinmuxTest, UnittestsUpToDate) {
```
This doesn't use the fixture, so consider using `TEST` and moving it to a separate test suite.

silvestrst replied:
Removed.

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - autogenerate?
+#define PINMUX_REGISTER_WIDTH_IN_BYTES (PINMUX_PARAM_REG_WIDTH / 8)
```
Do we expect this to ever change?

silvestrst replied:
No, but I find defines are more descriptive than hard coded numbers, I guess it could be `sizeof(uint32_t)`.

silvestrst replied:
Changed to `sizeof(uint32_t)`

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+
+// TODO - autogenerate?
+#define PINMUX_INSEL_RESET_DEFAULT 0
+#define PINMUX_OUTSEL_RESET_DEFAULT 0x20820820
```
This does not match the values in the technical specification page. Shouldn't this be `0x2082082`? I guess the last `0` is a typo.

silvestrst replied:
Thank you! Done.

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Sets the peripheral to the reset defaults.
+ */
+void dif_pinmux_reset(const dif_pinmux_t *pinmux) {
```
Can you make this function `static` if you are not planning to make it public (could be useful)?

silvestrst replied:
Done.

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+  }
+}
+
+dif_pinmux_init_error_t dif_pinmux_init(mmio_region_t base_addr,
```
I'm not sure if there is a use case for reading pinmux registers but I just wanted to point out that if `init` fails because `PINMUX_REGEN` is already false, we won't be able to lock pinmux registers in bootrom and init a new struct after jumping to flash, which could be used to read config registers. We can also consider initializing `pinmux` while still skipping reset and returning `kDifPinmuxInitErrorRegistersLocked`.

silvestrst replied:
It makes sense in case we want to read the configuration, thanks for the suggestion.

silvestrst replied:
More discussion in your other comment:
https://github.com/lowRISC/opentitan/pull/1763#discussion_r406512212

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+    return kDifPinmuxInselSetErrorRegistersLocked;
+  }
+
+  if (mio_select < kDifPinmuxInselTieToZero ||
+      mio_select > kDifPinmuxInselPadLast) {
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
```
DIF-specific style guide says that you don't need to check enums if the function can operate on the entire range. Also, checking `mio_select` but not checking `peripheral_input` is inconsistent. Please also see my comments regarding the enums in the header file.

silvestrst replied:
Now they are uint32_t typedefs, so upper range should be checked.

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+    return kDifPinmuxOutselSetErrorRegistersLocked;
+  }
+
+  if (mio_input_pad < kDifPinmuxMioInPadFirst ||
+      mio_input_pad > kDifPinmuxMioInPadLast) {
+    return kDifPinmuxOutselSetErrorBadMio;
+  }
```
Same here.

silvestrst replied:
Same as above.

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+  mmio_region_nonatomic_set_field32(pinmux->base_addr, register_offset, field);
+
+  return kDifPinmuxOutselSetErrorNone;
+}
```
Shouldn't we have functions for reading the current configuration?

silvestrst replied:
Hmmm, I see your point, I guess it makes sense to return these value in bulk (all values of insel, all values of outsel).

Not sure about the API, we might need to expose `PINMUX_PERIPH_INSEL_MULTIREG_COUNT` and `PINMUX_PERIPH_OUTSEL_MULTIREG_COUNT` to allow caller create a buffer to hold all values. This buffer then can be passed into a `*get` function.

It might be worth waiting until we auto generate pinmux defines in `top_earlgrey`, all this should be easier then.

alphan commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,231 @@
+ * These select which Pad Control MIO output is connected to the requested
+ * peripheral input (MIO output to pinmux input: MIO -> peripheral).
+ */
+typedef enum dif_pinmux_insel {
+  kDifPinmuxInselTieToZero = 0,
+  kDifPinmuxInselTieToOne = 1,
+  kDifPinmuxInselPadFirst = 2,
+  kDifPinmuxInselPadLast = kDifPinmuxInselTieToOne + DIF_PINMUX_MIO_NUM,
+} dif_pinmux_insel_t;
+#define DIF_PINMUX_INSEL_NUM (kDifPinmuxInselPadLast + 1)
```
I'm not sure if it is a good idea to define this enum with gaps while the other enums below cover the entire range.

silvestrst replied:
This enum is no more.

alphan commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,231 @@
+/**
+ * MIO inputs (pinmux output to MIO input: peripheral -> MIO).
+ */
+typedef enum dif_pinmux_mio_in {
+  kDifPinmuxMioInPadFirst = 0,
+  kDifPinmuxMioInPadLast = DIF_PINMUX_MIO_NUM - 1,
+} dif_pinmux_mio_in_t;
+#define DIF_PINMUX_MIO_IN_NUM (kDifPinmuxMioInPadLast + 1)
```
Same here.

silvestrst replied:
Same as above.

alphan commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,231 @@
+ */
+typedef struct dif_pinmux {
+  mmio_region_t base_addr; /**< Pinmux instance base address. */
+  bool write_enabled;      /**< Configuration registers locked/unlocked. */
```
This is a hardware-backed state that shouldn't be here according to the DIF doc.

silvestrst replied:
Removed.

alphan commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+      peripheral_input % PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  uint32_t field_offset = field_index * PINMUX_PERIPH_INSEL_IN_FIELD_WIDTH;
+
+  // INSEL field masks are the same for all of the fields.
+  bitfield_field32_t field = {
+      .mask = PINMUX_PERIPH_INSEL0_IN0_MASK,
+      .index = field_offset,
+      .value = mio_select,
+  };
+  mmio_region_nonatomic_set_field32(pinmux->base_addr, register_offset, field);
```
This is really ergonomic and a pleasure to read! Thank you!

silvestrst replied:
Thanks :)

alphan commented on sw/device/lib/pinmux.c:
```diff
@@ -6,36 +6,40 @@
+    dif_pinmux_outsel_set(&pinmux0, mio_in_pad, peripheral_out);
+    ++peripheral_out;
+    ++mio_in_pad;
+  }
```
If we use separate functions for constants and peripheral/mio indices, we can perform these operations in a single loop.

silvestrst replied:
I didn't quite find any good way of doing it, and I think `pinmux.c` will probably be a relatively short-lived - eventually these operations will be called either directly in the `boot_rom`, or in a tock driver?

silvestrst commented:
For visibility: there is a PR open to add power saving features to Pinmux in #1999 , which when they land (possibly beforehand) should be reflected in this DIF.

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Sets the peripheral to the reset defaults.
+ */
+void dif_pinmux_reset(const dif_pinmux_t *pinmux) {
```
Done.

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+    return kDifPinmuxInitErrorInvalidArgument;
+  }
+
+  // Check whether the peripheral can be configured.
+  bool write_enabled = mmio_region_get_bit32(base_addr, PINMUX_REGEN_REG_OFFSET,
+                                             PINMUX_REGEN_WEN);
+  if (!write_enabled) {
+    return kDifPinmuxInitErrorRegistersLocked;
+  }
```
Good idea, done!

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+    return kDifPinmuxInselSetErrorRegistersLocked;
+  }
+
+  if (mio_select < kDifPinmuxInselTieToZero ||
+      mio_select > kDifPinmuxInselPadLast) {
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
```
Similar to plic these are now typedef`ed to uint32_t, so the upper check is needed, but there are no enum variants anymore.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+  uint32_t field_offset = field_index * PINMUX_PERIPH_INSEL_IN_FIELD_WIDTH;
+
+  // INSEL field masks are the same for all of the fields.
+  bitfield_field32_t field = {
+      .mask = PINMUX_PERIPH_INSEL0_IN0_MASK,
+      .index = field_offset,
+      .value = mio_select,
+  };
```
Personal preference, in this case it seems cleaner to me.

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
+
+  // Register offset based on a given peripheral input and MIO select.
+  uint32_t register_index =
+      peripheral_input / PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  ptrdiff_t register_offset = PINMUX_PERIPH_INSEL0_REG_OFFSET +
+                              (register_index * PINMUX_REGISTER_WIDTH_IN_BYTES);
+
+  uint32_t field_index =
+      peripheral_input % PINMUX_PERIPH_INSEL_IN_FIELDS_PER_REG;
+  uint32_t field_offset = field_index * PINMUX_PERIPH_INSEL_IN_FIELD_WIDTH;
```
Good point, done.

silvestrst commented on sw/device/lib/pinmux.c:
```diff
@@ -6,36 +6,40 @@
-                  32 / PINMUX_MIO_OUTSEL0_OUT1_OFFSET,
-                  PINMUX_MIO_OUTSEL0_OUT0_MASK,
-                  PINMUX_PERIPH_OUTSEL_IDX_OFFSET);
+  // Assign a peripherl outputs to MIO inputs (1:1, don't assign if peripheral
```
Done.

silvestrst commented on sw/device/lib/pinmux.c:
```diff
@@ -6,36 +6,40 @@
-                            uint32_t mask, uint32_t start_v) {
-  uint32_t reg_value = 0;
-  uint32_t reg_offset = 0;
+static dif_pinmux_t pinmux0;
```
Done.

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/pinmux.c:
```diff
@@ -6,36 +6,40 @@
+void pinmux_init(void) {
+  mmio_region_t base_addr = mmio_region_from_addr(PINMUX0_BASE_ADDR);
+  if (dif_pinmux_init(base_addr, &pinmux0) != kDifPinmuxInitErrorNone) {
+    // TODO - print debug output.
```
Done.

silvestrst commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,379 @@
+    };
+    for (const auto &reg : outsel_registers) {
+      // TODO - autogenerate?
+      EXPECT_WRITE32(reg, 0x20820820);
```
Done.

silvestrst commented on sw/device/lib/dif/dif_pinmux.c:
```diff
@@ -0,0 +1,147 @@
+  return kDifPinmuxRegistersLockErrorNone;
+}
+
+dif_pinmux_insel_set_error_t dif_pinmux_insel_set(
```
Done, slightly differently - but amongst the same lines.

silvestrst commented:
Addressed most of the comments, please see the discussion on the open ones. Note, that IP is changing and the DIF will be extended in future (**please see the PR description**).

silvestrst commented:
* Fixed CI issue:
```
/home/vsts/work/1/s/sw/device/lib/dif/dif_pinmux.c:39:5: error: initializer element is not constant
     kDifPinmuxMioOutLast + kDifPinmuxInselFirstIn;
```
* Started using unittest defines instead of DIF defines to set expectations.

lenary commented on sw/device/lib/dif/dif_pinmux.h:
```diff
@@ -0,0 +1,168 @@
+ * @param mio_in_select Padring MIO input to be connected to `peripheral_input`.
+ * @return `dif_pinmux_result_t`.
+ */
+dif_pinmux_result_t dif_pinmux_insel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_peripheral_in_t peripheral_in,
+    dif_pinmux_insel_t mio_in_select);
+
+/**
+ * Sets the connection between a Padring MIO output and peripheral output.
+ *
+ * `mio_out` can be connected to any available `peripheral_select` output.
+ *
+ * @param pinmux Pinmux state data.
+ * @param mio_out Padring MIO output.
+ * @param peripheral_out_select Peripheral output select.
+ * @return `dif_pinmux_result_t`.
+ */
+dif_pinmux_result_t dif_pinmux_outsel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_mio_out_t mio_out,
+    dif_pinmux_outsel_t peripheral_out_select);
```
As with padctrl, these should check if the peripheral is locked before writing, and if it is locked return an error.

silvestrst commented:
@lenary sounds good, I would like you to expand little bit more on the following comment:
> We don't think that init should reset the pinmux

Are there any specific use-cases that you foresee that apply only to this peripheral alone (as the behaviour diverges from other peripherals)?

lenary commented:
> > We don't think that init should reset the pinmux
> 
> Are there any specific use-cases that you foresee that apply only to this peripheral alone (as the behaviour diverges from the other peripherals (UART, PLIC, ...))?

When we discussed pinmux, we discussed padctrl at the same time, and came to the conclusion that neither of these should be reset on init, because it can cause problems generally in ASICs or in testing on the FPGA. After your question here, we've thought further about this and summarised our thoughts in #2366.

msfschaffner commented:
> > > We don't think that init should reset the pinmux
> > 
> > 
> > Are there any specific use-cases that you foresee that apply only to this peripheral alone (as the behaviour diverges from the other peripherals (UART, PLIC, ...))?
> 
> When we discussed pinmux, we discussed padctrl at the same time, and came to the conclusion that neither of these should be reset on init, because it can cause problems generally in ASICs or in testing on the FPGA. After your question here, we've thought further about this and summarised our thoughts in #2366.

This is a prudent choice.

Waking up from deep sleep is another scenario where resetting during init could cause problems.
The `pinmux` / `padctrl` live in the AON domain, and do not loose their state when performing a deep sleep cycle. Resetting these IPs "blindly" after waking up could cause various issues related to the periphery that is connected to the chip.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,246 @@
+
+  std::vector<Register> registers_ = {
+      {
+          PINMUX_PERIPH_INSEL0_REG_OFFSET, PINMUX_PERIPH_INSEL0_IN4_OFFSET,
```
Nit: designated initializer.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,246 @@
+  };
+
+  // Set multireg expectations, one field per call for every multireg register.
+  void ExpectSetTests(const std::vector<Register> &regs, uint32_t field_width,
```
If you switch away from `std::vector`, you can replace this with
```
template <typename Regs>
void ExpectSetTests(const Regs&, ...) { ... }
```

You could even do something fancy with CTRP but I don't think that would be a great idea.

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,246 @@
+  static constexpr uint32_t kExpectOutselLastOut =
+      kExpectOutselFirstOut + (PINMUX_PARAM_NMIOPERIPHOUT - 1);
+
+  struct Register {
+    ptrdiff_t offset;    // Register offset from the base.
+    uint8_t last_field;  // Last field offset in the register.
+  };
```
I feel like this struct isn't explained too well. I can tell what it does from reading `ExpectSetTests`, but could you add a comment on top of it describing what it's used for?

mcy commented on sw/device/tests/dif/dif_pinmux_test.cc:
```diff
@@ -0,0 +1,246 @@
+    EXPECT_EQ(registers_.size(), PINMUX_PERIPH_INSEL_MULTIREG_COUNT);
+  }
+
+  std::vector<Register> registers_ = {
```
You use `std::vector<T> foo{}` above but `std::vector<T> foo = {}`. Prefer the latter, because it avoids the other `std::vector<T>` ctors.

silvestrst commented:
Currently it won't be possible to do a meaningful sanity test on the Verilator for gpio/pinmux/padctrl components, please see #2707 .

lenary commented:
With a view to getting this merged (it was opened in March), let's put the sanity test in a different PR. Let's try to separate the "changes requested" into what needs to be done before merge, and what can be done in a follow-up PR.

This PR doesn't need to cover the wakeup functionality of the pinmux.

silvestrst commented:
Pinmux and PadControl HW are expected to undergo major changes, a new PR will be opened.

silvestrst commented:
Superseded by #5368
